plugins/nvme/fu-nvme-device.c

```diff
@@ -0,0 +1,406 @@
+{
+	GString *str;
+
+	g_return_val_if_fail (buf != NULL, NULL);
+	g_return_val_if_fail (addr_start < addr_end, NULL);
+
+	str = g_string_new_len (NULL, addr_end + addr_start + 1);
+	for (guint16 i = addr_start; i <= addr_end; i++) {
+		gchar tmp = (gchar) buf[i];
+		/* skip leading spaces */
+		if (g_ascii_isspace (tmp) && str->len == 0)
+			continue;
+		if (g_ascii_isprint (tmp))
+			g_string_append_c (str, tmp);
+	}
+
+	/* nothing found */
+	if (str->len == 0) {
+		g_string_free (str, TRUE);
+		return NULL;
+	}
+	return g_strchomp (g_string_free (str, FALSE));
+}
+
+static gchar *
+fu_nvme_device_get_guid_safe (const guint8 *buf, guint16 addr_start)
+{
+	efi_guid_t guid_tmp;
+	guint c_cnt = 0;
+	g_autofree char *guid = NULL;
+
+	/* check the GUID is plausible */
+	for (guint i = 0; i < 16; i++)
+		c_cnt += buf[addr_start + i];
+	if (c_cnt < 0xff)
+		return NULL;
+	memcpy (&guid_tmp, buf + addr_start, 16);
+	if (efi_guid_to_str (&guid_tmp, &guid) < 0)
+		return NULL;
+	return g_strdup (guid);
+}
+
+static gboolean
+fu_nvme_device_submit_admin_passthru (FuNvmeDevice *self, struct nvme_admin_cmd *cmd, GError **error)
+{
+	if (ioctl (self->fd, NVME_IOCTL_ADMIN_CMD, cmd) < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed to issue admin command 0x%02x: %s",
+			     cmd->opcode,
+			     strerror (errno));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_nvme_device_identify_ctrl (FuNvmeDevice *self, guint8 *data, GError **error)
+{
+	struct nvme_admin_cmd cmd = {
+		.opcode		= 0x06,
+		.nsid		= 0x00,
+		.addr		= 0x0, /* memory address of data */
+		.data_len	= FU_NVME_ID_CTRL_SIZE,
+		.cdw10		= 0x01,
+		.cdw11		= 0x00,
+	};
+	memcpy (&cmd.addr, &data, sizeof (gpointer));
+	return fu_nvme_device_submit_admin_passthru (self, &cmd, error);
+}
+
+static gboolean
+fu_nvme_device_fw_commit (FuNvmeDevice *self,
+			  guint8 slot,
+			  guint8 action,
+			  guint8 bpid,
+			  GError **error)
+{
+	struct nvme_admin_cmd cmd = {
+		.opcode		= 0x10,
+		.cdw10		= (bpid << 31) | (action << 3) | slot,
+	};
+	return fu_nvme_device_submit_admin_passthru (self, &cmd, error);
+}
+
+static gboolean
+fu_nvme_device_fw_download (FuNvmeDevice *self,
+			    guint32 addr,
+			    const guint8 *data,
+			    guint32 data_sz,
+			    GError **error)
+{
+	struct nvme_admin_cmd cmd = {
+		.opcode		= 0x11,
+		.addr		= 0x0, /* memory address of data */
+		.data_len	= data_sz,
+		.cdw10		= (data_sz >> 2) - 1,
+		.cdw11		= addr >> 2,
+	};
+	memcpy (&cmd.addr, &data, sizeof (gpointer));
+	return fu_nvme_device_submit_admin_passthru (self, &cmd, error);
+}
+
+static void
+fu_nvme_device_parse_cns_maybe_dell (FuNvmeDevice *self, const guint8 *buf)
+{
+	g_autofree gchar *component_id = NULL;
+	g_autofree gchar *guid_efi = NULL;
+	g_autofree gchar *guid_id = NULL;
+
+	/* add extra component ID if set */
+	component_id = fu_nvme_device_get_string_safe (buf, 0xc36, 0xc3d);
```

Comment:
The start offset is right, but this should be 8 long not 7 long.

Reply:
Yup, it is. It's "inclusive" --hence the ` + 1` in the fu_nvme_device_get_string_safe() length calculation. I did it that way to match the constants in the NVMe spec. If it's confusing I can add a comment.

Reply:
Ah yes, a comment would be good.

Reply:
Done.